gazebo_ros/src/gazebo_ros_effort.cpp

```diff
@@ -0,0 +1,451 @@
+    "apply_body_wrench", std::bind(&GazeboRosEffortPrivate::ApplyBodyWrench, this,
+    std::placeholders::_1, std::placeholders::_2));
+
+  clear_body_wrenches_service_ = ros_node_->create_service<gazebo_msgs::srv::BodyRequest>(
+    "clear_body_wrenches", std::bind(&GazeboRosEffortPrivate::ClearBodyWrenches, this,
+    std::placeholders::_1, std::placeholders::_2));
+
+  apply_joint_effort_service_ = ros_node_->create_service<gazebo_msgs::srv::ApplyJointEffort>(
+    "apply_joint_effort", std::bind(&GazeboRosEffortPrivate::ApplyJointEffort, this,
+    std::placeholders::_1, std::placeholders::_2));
+
+  clear_joint_forces_service_ = ros_node_->create_service<gazebo_msgs::srv::JointRequest>(
+    "clear_joint_forces", std::bind(&GazeboRosEffortPrivate::ClearJointForces, this,
+    std::placeholders::_1, std::placeholders::_2));
+}
+
+void GazeboRosEffortPrivate::BodyWrenchTaskExecutor(const gazebo::common::UpdateInfo & _info)
+{
+  std::lock_guard<std::mutex> scoped_lock(wrench_lock_);
+
+  for (auto body_wrench_task = body_wrench_tasks_.begin();
+    body_wrench_task != body_wrench_tasks_.end(); ++body_wrench_task)
+  {
+    // check times and apply wrench if necessary
+    rclcpp::Time sim_time = Convert<builtin_interfaces::msg::Time>(_info.simTime);
+
+    auto start = (*body_wrench_task)->start_time;
+    auto duration = (*body_wrench_task)->duration;
+    auto end = start + duration;
+
+    if (!(*body_wrench_task)->link) {
+      body_wrench_tasks_.erase(body_wrench_task--);
+      RCLCPP_ERROR(ros_node_->get_logger(),
+        "Link [%s] does not exist. Deleting task", (*body_wrench_task)->link->GetName().c_str());
+    }
+
+    if (sim_time < start) {
+      continue;
+    }
+
+    if (sim_time <= end || duration.seconds() < 0.0) {
+      (*body_wrench_task)->link->SetForce((*body_wrench_task)->force);
+      (*body_wrench_task)->link->SetTorque((*body_wrench_task)->torque);
+    } else {
+      // remove from queue once expires
+      body_wrench_tasks_.erase(body_wrench_task--);
+    }
+  }
+}
+
+void GazeboRosEffortPrivate::JointEffortTaskExecutor(const gazebo::common::UpdateInfo & _info)
+{
+  std::lock_guard<std::mutex> scoped_lock(effort_lock_);
+
+  for (auto joint_effort_task = joint_effort_tasks_.begin();
+    joint_effort_task != joint_effort_tasks_.end(); ++joint_effort_task)
+  {
+    // check times and apply force if necessary
+    rclcpp::Time sim_time = Convert<builtin_interfaces::msg::Time>(_info.simTime);
+
+    auto start = (*joint_effort_task)->start_time;
+    auto duration = (*joint_effort_task)->duration;
+    auto end = start + duration;
+
+    if (!(*joint_effort_task)->joint) {
+      joint_effort_tasks_.erase(joint_effort_task--);
+      RCLCPP_ERROR(ros_node_->get_logger(),
+        "Joint [%s] does not exist. Deleting task", (*joint_effort_task)->joint->GetName().c_str());
+    }
+
+    if (sim_time < start) {
+      continue;
+    }
+
+    if (sim_time <= end || duration.seconds() < 0.0) {
```

Comment:
We should support durations that are less than 1 second, so we need to check nanoseconds too

Reply:
I don't think nanonseconds would make a difference here. Here we are just checking if the duration is negative. Only `seconds` part of the duration can be negative. `Nanoseconds` cannot be negative as it is `uint32_t` type.

Reply:
Oh you're right. Nanoseconds are only important when comparing `sim_time` and `end`.